jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java

```diff
@@ -35,31 +38,70 @@
+     * </p>
+     * <p>
+     * Servlet context attributes can be set with {@link #WEBSOCKET_BUFFER_POOL_ATTRIBUTE}, {@link #WEBSOCKET_INFLATER_POOL_ATTRIBUTE}
+     * and {@link #WEBSOCKET_DEFLATER_POOL_ATTRIBUTE} to override the {@link ByteBufferPool}, {@link DeflaterPool} or
+     * {@link InflaterPool} used by the components, otherwise this will try to use the pools shared on the {@link Server}.
+     * </p>
+     * @param server the server.
+     * @param servletContext the ServletContext.
+     * @return the WebSocketComponents that was created or found on the ServletContext.
+     */
     public static WebSocketComponents ensureWebSocketComponents(Server server, ServletContext servletContext)
     {
-        WebSocketComponents components = server.getBean(WebSocketComponents.class);
-        if (components == null)
-        {
-            InflaterPool inflaterPool = (InflaterPool)servletContext.getAttribute(WEBSOCKET_INFLATER_POOL_ATTRIBUTE);
-            if (inflaterPool == null)
-                inflaterPool = InflaterPool.ensurePool(server);
+        WebSocketComponents components = (WebSocketComponents)servletContext.getAttribute(WEBSOCKET_COMPONENTS_ATTRIBUTE);
+        if (components != null)
+            return components;
 
-            DeflaterPool deflaterPool = (DeflaterPool)servletContext.getAttribute(WEBSOCKET_DEFLATER_POOL_ATTRIBUTE);
-            if (deflaterPool == null)
-                deflaterPool = DeflaterPool.ensurePool(server);
+        InflaterPool inflaterPool = (InflaterPool)servletContext.getAttribute(WEBSOCKET_INFLATER_POOL_ATTRIBUTE);
+        if (inflaterPool == null)
+            inflaterPool = InflaterPool.ensurePool(server);
 
-            components = new WebSocketServerComponents(inflaterPool, deflaterPool);
-            server.addBean(components);
-        }
+        DeflaterPool deflaterPool = (DeflaterPool)servletContext.getAttribute(WEBSOCKET_DEFLATER_POOL_ATTRIBUTE);
+        if (deflaterPool == null)
+            deflaterPool = DeflaterPool.ensurePool(server);
 
-        servletContext.setAttribute(WEBSOCKET_COMPONENTS_ATTRIBUTE, components);
-        return components;
+        ByteBufferPool bufferPool = (ByteBufferPool)servletContext.getAttribute(WEBSOCKET_BUFFER_POOL_ATTRIBUTE);
+        if (bufferPool == null)
+            bufferPool = server.getBean(ByteBufferPool.class);
+
+        WebSocketComponents serverComponents = new WebSocketServerComponents(inflaterPool, deflaterPool, bufferPool);
+
+        // These components may be managed by the server but not yet started.
+        // In this case we don't want them to be managed by the components as well.
+        if (server.isManaged(inflaterPool))
+            serverComponents.unmanage(inflaterPool);
+        if (server.isManaged(deflaterPool))
+            serverComponents.unmanage(deflaterPool);
+        if (server.isManaged(bufferPool))
+            serverComponents.unmanage(bufferPool);
```

Comment:
We've very rarely, if ever needed to do this kind of explicit management of the management of beans.

Typically if a server bean is shared with handlers, it is added as auto in all places and the fact that the server starts it first is enough to make it managed there and not managed in the handlers.

Is this an issue here because the component is lazily added to the server and thus the order of initialisation is unreliable? 
 
Also, is it not the case that if the component is on the server, then it should not be managed here?
Ie, even if the components are created by a context, if they are put on the server then they can be managed by the server.

So at the very least, we need some good comments/doco describing the various options for the lifecycles and sharing of these components.

Reply:
> Is this an issue here because the component is lazily added to the server and thus the order of initialisation is unreliable?

This is an issue because the server may start the `ServletContextHandler` before it starts its `InflaterPool` / `DeflaterPool`. As they are all beans on the server it depends on what order they're added to what order they are started, and we create the `WebSocketServerComponents` in the `ServletContainerInitializer` when the `ServletContextHandler` is being started.

So without this change, this would work:
```java
server.addManaged(inflaterPool);
server.addManaged(deflaterPool);
server.setHandler(contextHandler);
```
But this would not work:
```java
server.setHandler(contextHandler);
server.addManaged(inflaterPool);
server.addManaged(deflaterPool);
```

> Also, is it not the case that if the component is on the server, then it should not be managed here?

The `WebSocketComponents` is now never taken from the server. One is always created for each `ServletContext`.